dss/storage/validator.py

```diff
@@ -1,23 +1,106 @@
+                                        )
+
+
+class SizeLimitError(IOError):
+    def __init__(self, url, limit) -> None:
+        super().__init__(f"{url} not cached. The URL's contents have exceeded {limit} bytes.")
+
+
+class S3UrlCache:
+    """Caches content of arbitrary URLs the first time they are requested. Currently only supports content lengths of up
+     to a few megabytes."""
+    _max_size_default = 64 * 1024 * 1024  # The default max_size per URL = 64 MB
+    _chunk_size_default = 1024 * 1024  # The default chunk_size = 1 MB
+
+    def __init__(self, logger: logging.Logger,
+                 max_size: int = _max_size_default,
+                 chunk_size: int = _chunk_size_default):
+        """
+        :param max_size: The maximum number of bytes of content that can be cached. An error is returned if the content
+        is greater than max_size, and the URL is not cached.
+        :param chunk_size: The amount of content retrieved from the URL per request."""
+        self.max_size = max_size if max_size is not None else self._max_size_default
+        self.chunk_size = chunk_size if chunk_size is not None else self._chunk_size_default
+        self.s3_client = boto3.client('s3')
+
+        # A URL's contents are only stored in s3 to keep the data closer to the aws lambas which use them.
+        self.blobstore, _, self.bucket = Config.get_cloud_specific_handles(Replica.aws)
+        self.logger = logger
 
-def read_json_url(link):
-    gcontext = ssl.SSLContext(ssl.PROTOCOL_TLSv1)
-    with urlopen(link, context=gcontext) as f:
-        return json.load(f)
+    def resolve(self, url: str) -> bytearray:
+        """
+        Requests the contents of a URL by first checking for the contents in an s3 bucket. If not present, the contents
+        are retrieved from the URL and stored in s3.
 
+        :param url: The JSON data from which to remove fields.
+        """
+        key = self._create_key(url)
 
-def remove_json_fields(json_data: dict, path: List[str], fields: List[str]):
+        # if key in s3 bucket return value from there.
+        try:
+            content = bytearray(self.blobstore.get(self.bucket, key))
+        except BlobNotFoundError:
+            with requests.get(url, stream=True) as resp_obj:
+                content = bytearray()
+                for chunk in resp_obj.iter_content(chunk_size=self.chunk_size):
+                    #  check if max_size exceeded before storing content to avoid storing large chunks
+                    if len(content) + len(chunk) > self.max_size:
+                        raise SizeLimitError(url, self.max_size)
+                    content.extend(chunk)
+            self._upload_content(key, url, content)
+            self.logger.info("%s", f"{url} not found in cache. Added it to {self.bucket} with key {key}.")
+        return content
+
+    def invalidate(self, url):
+        self.blobstore.delete(self.bucket, self._create_key(url))
```

Comment:
This will raise an exception if the URL isn't currently cached. I think it shouldn't, i.e. this method should be idempotent. The important post condition is that the cache does not contain the URL and this function should not fail if that post condition is already satisfied. Don't forget to add a test for idempotence.

And while I asked proposed the name `invalidate` in an earlier review, I think `evict` would be a better name. I got the two confused. Eviction is the term commonly used.

https://www.google.com/search?q=cache+eviction

Reply:
Consider adding a doc string.

Reply:
blobstore returns False if nothing was deleted or something else if something was deleted. No exception should be raised.

Reply:
added a doc string